exporters/trace/src/main/java/com/google/cloud/opentelemetry/trace/TraceConfiguration.java

```diff
@@ -231,17 +240,44 @@ public final Builder addAttributeMapping(String otelKey, String gcpKey) {
      * @return a {@code TraceConfiguration}.
      */
     public TraceConfiguration build() {
+      Supplier<String> projectIdProvider = generateProjectIdProvider();
       // Make a defensive copy of fixed attributes.
       setFixedAttributes(Collections.unmodifiableMap(new LinkedHashMap<>(getFixedAttributes())));
-      Preconditions.checkArgument(
-          !Strings.isNullOrEmpty(getProjectId()),
-          "Cannot find a project ID from either configurations or application default.");
       for (Map.Entry<String, AttributeValue> fixedAttribute : getFixedAttributes().entrySet()) {
         Preconditions.checkNotNull(fixedAttribute.getKey(), "attribute key");
         Preconditions.checkNotNull(fixedAttribute.getValue(), "attribute value");
       }
       Preconditions.checkArgument(getDeadline().compareTo(ZERO) > 0, "Deadline must be positive.");
-      return autoBuild();
+      TraceConfiguration builtTraceConfiguration = autoBuild();
+      builtTraceConfiguration.setProjectIdProvider(projectIdProvider);
+      return builtTraceConfiguration;
+    }
+
+    @Nonnull
+    private Supplier<String> generateProjectIdProvider() {
+      String userProvidedProjectId = getUserProvidedProjectId();
+      if (Strings.isNullOrEmpty(userProvidedProjectId)) {
+        return () -> {
```

Comment:
Should this be memoized ?

Currently this is only being called once.

Since the Supplier generated can only be retrieved by a package-private getter, it is not exposed to the end user and we have control on how often this used. 
`get()` on this supplier is called during the initialization of trace exporter. So per `TraceExporter` it will be called once. 

Thoughts @punya @jsuereth ?

Reply:
If you follow my suggestions from the other comment, you'll end up memoizing but in a slightly different place.